### DISCUSSION TITLE
Button is not rendering children

### DISCUSSION THREAD

**Discussion by mg-aceik** [mantinedev/mantine#4972] [Q&A] [answered]
I am using a Link component (source is [here](https://github.com/Sitecore/jss/blob/dev/packages/sitecore-jss-nextjs/src/components/Link.tsx)) which has a bit of logic to it.
I want to use the Mantine Button with it.

Here is my test, they are close (very similar) but both of them do not render the actual button label.

        import {
          Link as JssLink,
        } from '@sitecore-jss/sitecore-jss-nextjs';
        import {
          Overlay,
          Button,
        } from '@mantine/core';

        <JssLink field={props.fields.PromoLink} />

        <Button
          component={JssLink}
          field={props.fields.PromoLink}
          size="xl"
          radius="xl"
          className={heroBackroundImageClasses.control}
        />

        <Button
          renderRoot={({ ...others }) => <JssLink field={props.fields.PromoLink} {...others} />}
          size="xl"
          radius="xl"
          className={heroBackroundImageClasses.control}
        />

Rendered output:

        <a title="" target="" class="" href="/Contact-Us">Get Started</a>
        <a title="" target="" class="mantine-focus-auto mantine-active Promo_HeroBackroundImage_control__bcJBO m-77c9d27d mantine-Button-root m-87cf2631 mantine-UnstyledButton-root" href="/Contact-Us" data-size="xl" style="--button-height: var(--button-height-xl); --button-padding-x: var(--button-padding-x-xl); --button-fz: var(--mantine-font-size-xl); --button-radius: var(--mantine-radius-xl);"><span class="m-80f1301b mantine-Button-inner"><span class="m-811560b9 mantine-Button-label"></span></span></a>
        <a title="" target="" class="mantine-focus-auto mantine-active Promo_HeroBackroundImage_control__bcJBO m-77c9d27d mantine-Button-root m-87cf2631 mantine-UnstyledButton-root" data-size="xl" type="button" href="/Contact-Us" style="--button-height: var(--button-height-xl); --button-padding-x: var(--button-padding-x-xl); --button-fz: var(--mantine-font-size-xl); --button-radius: var(--mantine-radius-xl);"><span class="m-80f1301b mantine-Button-inner"><span class="m-811560b9 mantine-Button-label"></span></span></a>

Any help would be appreciated.

**Comment by rtivital** [comment]
In your example, Button does not render any children because you did not set any children.

**Reply by mg-aceik** [reply]
I didn't set any children in JssLink either and it is rendering the label - "Get Started"

**Reply by rtivital** [reply]
It is hard to tell what is wrong. If you do not use generic components, use `component` prop instead of `renderRoot`.

**Reply by mg-aceik** [reply]
I'm using component prop in this example.
I'll try set up a repro.

**Reply by rtivital** [reply]
You can find examples with renderRoot and component props here – https://codesandbox.io/s/suspicious-kalam-x9ym47?file=/src/App.tsx

Both cases work correctly with regular `a` tag. The issue is with your JssLink component.

**Reply by mg-aceik** [reply]
Repro is here - https://codesandbox.io/s/distracted-hertz-5p6ngl?file=/src/App.tsx

**Reply by mg-aceik** [reply]
This JssLink component is from a CMS so I don't have control of it. The source is here https://github.com/Sitecore/jss/blob/dev/packages/sitecore-jss-nextjs/src/components/Link.tsx

**Reply by rtivital** [reply]
JssLink component does not support `children` prop – you cannot render markup inside it. Event if are able to render it (https://codesandbox.io/s/withered-hooks-j3q6f9), it will not have correct styles because extra elements are not passed to the children.
`children` is a requirement for `renderRoot` and `component` props to work correctly.

**Reply by mg-aceik** [reply]
An example with children prop - https://codesandbox.io/s/peaceful-danny-549xvz?file=/src/App.tsx

**Reply by rtivital** [reply]
It is the same example as I've provided and explained why it does not work

**Reply by mg-aceik** [reply]
That JssLink calls down to this [Link component](https://github.com/Sitecore/jss/blob/76dc8a54ba77baf6dec26aa7e05bede41747e9c5/packages/sitecore-jss-react/src/components/Link.tsx#L114)
Essentially: 
 
    const linkText = showLinkTextWithChildrenPresent || !children ? link.text || link.href : null;

    resultTags.push(
      React.createElement(
        'a',
        { ...anchorAttrs, ...otherProps, key: 'link', ref },
        linkText,
        children
      )
    );

    return <React.Fragment>{resultTags}</React.Fragment>;

What would you need to change to get this to work as expected?

**Reply by rtivital** [reply]
The Link component must accept `children` prop with any content. On your side, you must pass content of the Button to `children` instead of `field` prop.

**Reply by mg-aceik** [reply]
Apologies, but it looks to me as if it is handling that scenario already?

**Reply by rtivital** [reply]
See example – https://codesandbox.io/s/withered-hooks-j3q6f9?file=/src/App.tsx:567-575

**Reply by mg-aceik** [reply, marked as answer]
Ok thanks, bit disappointing to need that workaround. In the end I have gone with this solution of a custom Button wrapper:

    import {
      Link as JssLink,
      LinkField as JssLinkField,
      LinkFieldValue as JssLinkFieldValue,
    } from '@sitecore-jss/sitecore-jss-nextjs';
    import { forwardRef } from 'react';
    import { createPolymorphicComponent, Button, ButtonProps } from '@mantine/core';

    interface CustomButtonProps extends ButtonProps {
      field: JssLinkField | JssLinkFieldValue;
    }

    export default createPolymorphicComponent<'a', CustomButtonProps>(
      // eslint-disable-next-line react/display-name
      forwardRef<HTMLAnchorElement, CustomButtonProps>(
        (props: CustomButtonProps, ref): JSX.Element | null => {
          const { field } = props;
          const value = (
            (field as JssLinkFieldValue).href ? field : (field as JssLinkField).value
          ) as JssLinkFieldValue;
          return (
            <Button component={JssLink} {...props} ref={ref}>
              {value.text}
            </Button>
          );
        }
      )
    );
